File: client/client/Classes/Frontend/Behaviour/State/IconJumpState.m
// ======================================================================
// Project Name    : ios_arkit
//
// Copyright © 2018 U-CREATES. All rights reserved.
//
// This source code is the property of U-CREATES.
// If such findings are accepted at any time.
// We hope the tips and helpful in developing.
// ======================================================================
#import <iOSObjectiveCFoundation/iOSObjectiveCFoundation.h>
#import "IconJumpState.h"
@implementation IconJumpState
- (id)init {
    self = [super init];
    self->parabora = [[Parabora alloc] init];
    return self;
}
- (void)create {
    self->parabora.velocity = 1.0f;
    self->parabora.mass = 1.5f;
    self->parabora.radian = [Angle toRadian:90.0f];
    [self.timeLine restore];
    self.timeLine.rate = 0.1f;
    return;
}
- (void)update:(NSTimeInterval)delta {
    SCNVector3 position = [self->parabora create:self.timeLine.currentFrame];
    if (0.0f > position.y) {
        self.owner.node.position = self.owner.defaultPosition;
        Notifier* notifier = [Notifier getInstance];
        [notifier notify:OnActionComplete];
        self.complete = YES;
        return;
    } else {
        float nx = self.owner.node.position.x;
        float ny = position.y;
        float nz = self.owner.node.position.z;
        self.owner.node.position = SCNVector3Make(nx, ny, nz);
    }
    [self.timeLine next:delta];
    return;
}
@end

File: client/iOSSwiftClient/Classes/Frontend/Behaviour/State/IconJumpState.swift
// ======================================================================
// Project Name    : ios_arkit
//
// Copyright © 2018 U-CREATES. All rights reserved.
//
// This source code is the property of U-CREATES.
// If such findings are accepted at any time.
// We hope the tips and helpful in developing.
// ======================================================================
import SceneKit
import iOSSwiftFoundation
class IconJumpState: FiniteState<IconBehaviour> {
    var parabora: Parabora!
    open override func create() -> Void {
        self.parabora = Parabora()
        self.parabora.velocity = 1.0
        self.parabora.mass = 1.0
        self.parabora.radian = Angle.toRadian(degree: 90.0)
        self.timeLine.restore()
        self.timeLine.rate = 0.1
        return
    }
    open override func update(delta: TimeInterval) -> Void {
        let position: SCNVector3 = self.parabora.create(currentFrame: self.timeLine.currentFrame)
        if (0.0 > position.y) {
            self.owner.node.position = self.owner.defaultPosition
            let notifier: Notifier = Notifier.getInstance()
            notifier.notify(message: NotifyMessage.OnActionComplete)
            self.complete = true
            return
        } else {
            let nx: Float = self.owner.node.position.x
            let ny: Float = position.y
            let nz: Float = self.owner.node.position.z
            self.owner.node.position = SCNVector3Make(nx, ny, nz)
        }
        self.timeLine.next(delta: delta)
        return
    }
}
